Add async F* IDE API

klinvill opened the pull request:
Resolves #27. Proposed in place of the unmerged suffix of #20. This PR adds an async API to communicate with F* over the IDE protocol. Specifically, it provides async methods that can be called to send requests to F*. In the case that a response is expected, the response can be accessed via the returned promise.

This PR is still a Work-In-Progress. There are many outstanding TODOs and changes that need to be made depending on the exact behavior of the F* IDE protocol. For example, the proof-status and IdeDiagnostic messages sent by F* are currently not handled anywhere.

gebner commented on server/src/fstar_connection.ts:
```diff
@@ -0,0 +1,512 @@
+			// TODO(klinvill): should we throw exceptions here or reject the
+			// returned promise? Although I believe the semantics of async
+			// functions are such that this will automatically result in a
+			// rejected promise?
```
Indeed, this will just reject the promise returned by `request` (which is imho the right behavior).  You can easily try this out in the REPL:
```
$ node
Welcome to Node.js v21.5.0.
Type ".help" for more information.
> (async () => { throw Error('lets try it out') })()
Promise {
  <rejected> Error: lets try it out
      at REPL2:1:22
      at REPL2:1:49
      at ContextifyScript.runInThisContext (node:vm:121:12)
      at REPLServer.defaultEval (node:repl:599:22)
      at bound (node:domain:432:15)
      at REPLServer.runBound [as eval] (node:domain:443:12)
      at REPLServer.onLine (node:repl:929:10)
      at REPLServer.emit (node:events:531:35)
      at REPLServer.emit (node:domain:488:12)
      at [_onLine] [as _onLine] (node:internal/readline/interface:416:12),
  [Symbol(async_id_symbol)]: 268,
  [Symbol(trigger_async_id_symbol)]: 6
}
```

gebner commented on server/src/fstar_connection.ts:
```diff
@@ -0,0 +1,512 @@
+			throw new Error(msg);
+		}
+
+		const internal_promise =  withResolvers();
+		if (expect_response) {
+			// Only keep track of responses if we expect a response
+			this.pending_responses.set(qid, {ip: internal_promise, is_stream});
+		}
+		// TODO(klinvill): Should we always return a promise even if no response
+		// is expected? In that case the promise is just ignored.
+		return internal_promise.promise;
```
I think it would be shorter and cleaner to write this without the additional wrappers:
```suggestion
		if (expect_response) {
		        return new Promise((resolve, reject) =>
		                this.pending_responses.set(qid, {resolve, reject, is_stream}));               
		}
```

> Should we always return a promise even if no response is expected?

If we stored it in `this.pending_responses`, then that would lead to a memory leak (since we'll never receive a response that would delete it).

gebner replied:
> Should we always return a promise even if no response is expected?

Sorry, I think I misunderstood the question here.

I believe what you're curious about is the following:
```js
async sendNotification() {
  return new Promise((resolve, reject) => {
    /* discard the resolve handler because we don't expect a response */
  });
}
```

This is not a good idea.  When you call `await sendNotification()`, the caller will then wait forever because the promise is not resolved.

gebner commented on server/src/fstar_connection.ts:
```diff
@@ -0,0 +1,512 @@
+	//
+	// @param expect_response indicates if a response from F* is expected for
+	// this request.
+	private async request(msg: any, debug: boolean, expect_response = true, is_stream = false): Promise<any> {
```
```suggestion
	private async request(msg: any, expect_response = true, is_stream = false): Promise<any> {
```

`debug` should probably be a field on `FStarConnection`.  I don't see a reason why we would want to have different debugging settings per request.

klinvill replied:
The reason is that the debugging settings may change between requests since they're set at an editor level, rather than at a connection level. I suppose we could have debug as the field on the connection anyway and modify the `updateConfigurationSettings()` logic on the server to always set the debug field with the new value for each connection for each documentState. I'll go ahead and implement that change.

gebner commented on server/src/fstar_connection.ts:
```diff
@@ -0,0 +1,512 @@
+		const expect_response = true;
+		const is_stream = true;
+		return this.request(query, debug, expect_response, is_stream);
+	}
```
I'd factor out the common parts of the two functions here:

```suggestion
        async fullBufferQuery(query: FullBufferQuery): partialResult<IdeProgress> {
		if (!this.fstar.supportsFullBuffer) {
			throw new Error("ERROR: F* process does not support full-buffer queries");
		}
		const expect_response = true;
		const is_stream = true;
		return this.request(query, expect_response, is_stream);
        }

	async fullBufferRequest(code: string, kind: 'full' | 'lax' | 'cache' | 'reload-deps', withSymbols: boolean): partialResult<IdeProgress> {
		return this.fullBufferQuery({
			query: "full-buffer",
			args: {
				kind,
				"with-symbols": withSymbols,
				code: code,
				line: 0,
				column: 0,
			},
		});
	}
```

gebner commented on server/src/fstar_connection.ts:
```diff
@@ -0,0 +1,512 @@
+
+		const expect_response = false;
+		this.request(query, debug, expect_response);
+	}
```
I would introduce a separate function for simple requests that have exactly one response, to reduce the Boolean blindness here.

```suggestion
        async simpleRequest<T>(query: T) {
                const expectResponse = true;
                return this.request(query, expectResponse);
        }

	// Request to map a file into F*'s virtual file system.
	//
	// For more details, see:
	// https://github.com/FStarLang/FStar/wiki/Editor-support-for-F*#vfs-add
	async vfsAddRequest(filePath: string, contents: string, debug: boolean) {
	        return this.simpleRequest<VfsAdd>({
			query: "vfs-add",
			args: {
				filename: filePath,
				contents: contents,
			},
		});
	}
```

klinvill commented:
@gebner what are your thoughts on the direction of this PR? Does this mesh with how you envisioned the async API?

@gebner or @nikswamy I have a bunch of questions about the F* IDE protocol that I think are required to get this PR working as intended. Specifically:

- What are the semantics of the `cancel` request? From Gabriel's comment (https://github.com/FStarLang/fstar-vscode-assistant/issues/27#issuecomment-1890185735) it seems like cancel requests are ignored unless a full-buffer query is being processed. From comments in the LSP server code, it mentions that `A Cancel message should be sent to F* when the document changes at a given range, to stop it from verifying the part of the buffer that has changed`. Are the semantics then that if a full-buffer query is running, the full-buffer query will not process any part of the spec beyond the requested cancel line and column? Can it interrupt the current fragment that's in the middle of being checked, or does it only affect further fragments? Furthermore, is there an expected response from `cancel` requests?
- Is there any expected response from `restart-solver` requests?
- Are full-buffer requests the only requests that may receive multiple responses?
- What does the `query-id` in the response messages mean? I would expect it to match the relevant query, but it looks like responses to full-buffer requests can contain fractional query-ids. For example, in response to full-buffer query-id `2`, I see response messages with query-ids `2`, `2.1`, `2.2`, etc. These seem to be monotonically increasing, but there are multiple responses with query-id `2.1` so they don't seem to be monotonically strictly increasing. Are the fractional parts semantically meaningful?
- What requests can result in `proof-state` status messages? I'm assuming `full-buffer` requests, but are there other requests that can prompt these messages too? Is there always a terminating message (like `full-buffer-finished`) that comes after these messages, or is it possible for a `proof-state` status message to be the last response to a query?
- What requests can result in `symbol` responses? I thought these would be in response to `lookup` requests, but I see some `symbol` responses coming back for `full-buffer` requests. Is there always a terminating message (like `full-buffer-finished`) that comes after these messages, or is it possible for a `symbol` response to be the last response to a query?
-  What requests can result in an `error` response being sent back? There's a subset listed on https://github.com/FStarLang/FStar/wiki/Editor-support-for-F*#list-of-status-messages, but can an error response be sent back for any request? If an error response is sent back for a query that normally receives several response messages, is there always a terminating message (like `full-buffer-finished`) that comes after an error response, or is it possible for a error response to be the last response to a query?

I also have a few other questions about intended behavior including:

- `restartSolver` only restarts the solver for the normal F* process, not the lax one. Is this intended behavior?
- The formatting calls to F* spin up a fresh process for that request. Is there a reason a fresh process would be needed instead of re-using either the normal or lax F* processes?

gebner commented on server/src/server.ts:
```diff
@@ -588,16 +520,14 @@ export class Server {
-			this.sendRequestForDocument(doc, autoCompletionRequest, this.configurationSettings.flyCheck ? "lax" : undefined);
+			const lax = this.configurationSettings.flyCheck ? "lax" : undefined;
+			const fstar_conn = this.getFStarConnection(doc, lax);
+			// TODO(klinvill): should we await the response here? The autocomplete response table is populated asynchronously.
```
Eventually we want to drop the response table entirely.  See #28.  Then we can just do `autoCompleteResponses = await fstar_conn.autocompleteRequest(...)` at the beginning of the function.

klinvill replied:
I'll just leave this for a future PR since it's already documented as an issue.

gebner commented on server/src/fstar_connection.ts:
```diff
@@ -0,0 +1,512 @@
+			// E.g. if the query id is 2, the responses seem to come back with
+			// query ids 2, 2.1, 2.2, 2.3, etc. To deal with this behavior I
+			// truncate the responses query-id. Is this the proper behavior for
+			// this case?
```
Yes, this is correct.

gebner commented on server/src/fstar_connection.ts:
```diff
@@ -0,0 +1,512 @@
+				} else if (responseType === 'error') {
+					// TODO(klinvill): Should we instead reject the promise
+					// here? Or pass along the error responses as the resolved
+					// promise value?
```
This is a good question.  For example, the lookup query returns an error if it cannot resolve the identifier.  That is not really something that I would use an exception for.  So I'd tend to have a `Promise<IdeQueryResponse | IdeError>` type.

klinvill replied:
I update the return types for the fullBufferQuery requests to have the type `Promise<IdeProgress | IdeErrors>`. Are there other methods that could result in IdeError responses being sent? On the protocol wiki page (https://github.com/FStarLang/FStar/wiki/Editor-support-for-F*#list-of-status-messages) I only see this being the case for `push` requests.

gebner commented on server/src/fstar_connection.ts:
```diff
@@ -0,0 +1,512 @@
+	// TODO(klinvill): previously the supportsFullBuffer was set for both the
+	// lax and non-lax F* processes when a single protocol info message was
+	// received. I changed the behavior here to only set it for the process the
+	// response comes from. Does this change break anything?
```
Sounds good to me.

gebner commented on server/src/fstar_connection.ts:
```diff
@@ -0,0 +1,512 @@
+				// TODO(klinvill): info messages are just logged and don't resolve the corresponding pending_responses promise. Is this the right behavior (that info messages are just extraneous and should be logged and ignored)?
+				console.log("Info: " + r.contents);
+			} else if (r.kind === "response" && r.response) {
+				const responseType = FStarConnection.decideIdeReponseType(r.response);
```
Is there a reason why we need to dispatch on the (heuristically guessed) response type here?  I think we do the same in all cases (i.e., resolve the promise associated with the query id).

klinvill replied:
Ah that was the case, but with my latest change I wrap IdeError responses with a `kind` field that can be used to easily distinguish them (instead of heuristically guessing based on the response being an array or instead of passing through the full `IdeQueryResponse` object). Do you think there's a better way to do this?

gebner replied:
The proper way to figure out if a response was a success or not is by looking at the `status` field.  But I would keep it simple and just pass the whole `msg` back to the caller.  Then we can check `resp.status == 'error'` etc. at the call site.

klinvill replied:
My latest commit now just fulfills every promise with the full `IdeQueryResponse` object. To keep some of the informative type information about expected responses, I opted to parameterize the `IdeQueryResponse` type by the types in its `response` and `content` fields. This allows me to specify that the return type for a `lookupQuery` should include an `IdeSymbol` in its response (e.g. `Promise<IdeQueryResponseR<IdeSymbol>>`), but I also leave the response type for full-buffer queries open to any response (e.g. `StreamedResult<AnyIdeQueryResponse>`).

Any thoughts on this attempt to balance informative types with many possible responses?

gebner replied:
I wouldn't constrain the type parameters so much:
```typescript
// A query response envelope
export interface IdeQueryResponse {
	'query-id': string;
	kind: 'protocol-info' | 'response' | 'message';
	status?: 'success' | 'failure';
	level?: 'progress' | 'proof-state' | 'info';
}

export interface IdeQueryResponseR<R> extends IdeQueryResponse {
	response: R;
}
export interface IdeQueryResponseC<C> extends IdeQueryResponse {
	contents: C;
}
```

With the above, you'll even know that the `response` field of `IdeQueryResponseR` is defined.

Another maybe even clearer option is to have one type for every response:
```typescript
export interface IdeQueryResponse {
	'query-id': string;
	status?: 'success' | 'failure';
}

export interface IdeAutoCompleteResponse extends IdeQueryResponse {
	response: [number, string, string][];
}

// ...
```

For these kind of interface descriptions, I find inheritance to be a lot more readable than genericity.  (In case it's not clear, in typescript the value `{'query-id': '5', foo: 42}` does have the type `IdeQueryResponse`.  You can always have additional properties that are not mentioned in the interface.)

klinvill replied:
Sounds good to me, I like the approach of having one type for every response. I went ahead and pushed up a commit that does this. The downside is there's now quite a few mostly trivial types (e.g. `IdeSymbolResponse` just extends `IdeQueryResponse` and adds the field `response: IdeSymbol`), but we can always refactor this later to move the details into the response type as desired (e.g. move the `IdeSymbol` definition into `IdeSymbolResponse`).

gebner replied:
I see what you mean.  I wouldn't worry too much about the details.  The important part is that the types check out.

gebner commented on server/src/fstar_connection.ts:
```diff
@@ -0,0 +1,512 @@
+// completed queries.
+//
+// TODO(klinvill): would be nice to have this be a more informative type than
+// potentially containing `undefined`. E.g. to have Continuing and Done types.
```
IMHO what you have right now is the idiomatic way to do it.

gebner commented on server/src/fstar_connection.ts:
```diff
@@ -0,0 +1,512 @@
+//
+// TODO(klinvill): would also be nice to have an iterator interface for
+// partialResults (or for a wrapper type like streamResults)
+export type partialResult<T> = Promise<[T, partialResult<T> | undefined]>;
```
```suggestion
export type StreamedResult<T> = Promise<[T, StreamedResult<T> | undefined]>;
```

Nitpick: usually types are capitalized in typescript.

gebner commented on server/src/fstar_connection.ts:
```diff
@@ -0,0 +1,512 @@
+		};
+
+		// TODO(klinvill): I believe there's no responses to restart-solver requests, is
+		// that correct?
```
Yes, this is correct.

gebner commented on server/src/fstar_connection.ts:
```diff
@@ -0,0 +1,490 @@
+	// @param expect_response indicates if a response from F* is expected for
+	// this request.
+	private async request(msg: any, expect_response = true, is_stream = false): Promise<any> {
+		// TODO(klinvill): do we need to worry about racing requests here that might result in the same query-id? That would cause issues when tracking responses.
```
Concurrency in javascript is a bit different than in C. :smile: 

Javascript is completely single threaded, and promises are just a nice API for cooperative multi-threading.  In other words, everything is atomic unless there's an await in the middle.

gebner commented on server/src/fstar_handlers.ts:
```diff
@@ -193,7 +32,10 @@ function handleIdeSymbol(textDocument: TextDocument, response: IdeSymbol, server
-function handleIdeProofState(textDocument: TextDocument, response: IdeProofState, server: Server) {
+//
+// TODO(klinvill): with my latest edits, this function isn't called anywhere.
+// Where should it be handled? server.ts? fstar_connection.ts?
```
I think it should be called by whoever sends the full buffer query, i.e., `validateFStarDocument`.  (And the proof state messages should be included in the full buffer response stream.)

Unfortunately the full buffer protocol is a bit special.

gebner commented on server/src/fstar_handlers.ts:
```diff
@@ -298,9 +140,13 @@ function handleIdeProgress(textDocument: TextDocument, contents: IdeProgress, la
+//
+// TODO(klinvill): with my latest edits, this function is only called for
+// fullBufferQuery requests. Should it also be called for other requests? What
+// about ones for which responses aren't expected?
```
Only full buffer requests produce messages that should be shown as diagnostics, so yes.

klinvill commented:
This latest version seems to work according to some ad-hoc testing I've done so I'm marking it as ready for review. However, I still have a slew of TODO comments in the code, and I think these changes could benefit from some further ad-hoc testing (unless further tests like from #26 are added and passed).

gebner commented on server/src/fstar_connection.ts:
```diff
@@ -0,0 +1,434 @@
+	debug: boolean;
+
+	constructor(fstar: FStar, debug: boolean) {
+		this.debug = debug;
```
```suggestion

	constructor(fstar: FStar, public debug: boolean) {
```

:smile:

klinvill replied:
Neat! I'm torn between the more verbose representation I originally had (since it visually lays out the debug field next to the other fields) and the more concise syntax proposed here. I went ahead with the concise version.

gebner commented:
Looking through the todos, there are understandably a lot of questions about the full buffer request and cancellation.  Let me try to explain the big picture to the best of my knowledge.

The purpose of the full buffer request is to process multiple declarations of an F* file.  It reports progress and warnings / error messages along the way.  Every time it is done with a declaration it checks if a cancel request came in, and if so, stops the full buffer request.  (And it literally only checks the next incoming request, so any other request that arrives in between will break cancellation.)

So conceptually, there are two states in the IDE protocol:
 1. Currently processing a full buffer request
 2. Open for other requests (like symbol lookups, etc.)

In state (1) the only supported request is `cancel`, and no other request should be sent.

Cancellation is completely ignored in state (2).

gebner commented:
I did some more testing and ran into several issues:

1. The F* server sends a response to the `vfs-add` request.
2. Unhandled exceptions seem to crash the LSP server.  I'm not sure why that happens.
3. The full buffer request can also respond with errors apparently:

```
Unhandled full-buffer response: {"kind":"message","query-id":"2","level":"error","contents":"Unexpected error; please file a bug report, ideally with a minimized version of the source program that triggered the error.\nFile \"fstar-lib/generated/FStar_Parser_AST_Util.ml\", line 651, characters 4-9: Pattern matching failed\n"}
```

gebner commented:
> Unhandled exceptions seem to crash the LSP server. I'm not sure why that happens.

I think that happens because we throw exceptions in promise that we never catch.  This is called "unhandled rejection".  It doesn't seem to cause any issues in the REPL, and in the past I've always gotten warnings for that, not errors, so I didn't think too much of it.

There's probably three things we should do:
1. Make unhandled rejections non-fatal by registering a handler:
```javascript
// Add this in server/main.ts
process.on('unhandledRejection', error => console.log('Unhandled rejection:', error));
```

2. Make sure to await every async function.  (In this case, the `this.handleFullBufferResponse` call in `validateFStarDocument`.)
3. Add a [linting rule](https://eslint.org/docs/latest/rules/require-await) to enforce (2).

klinvill commented:
> 1. The F* server sends a response to the `vfs-add` request.

This should be fixed now, I changed `vfsAddRequest` to return a promise that gets fulfilled with the response (which is then promptly ignored unless an exception occurs).

klinvill commented:
> 3\. The full buffer request can also respond with errors apparently:

Should these error responses result in an error that the IDE highlights? Or just logged to the console? It looks like there is file + line information, but it's buried in a string message that probably won't always occur (as opposed to the `IdeError` messages which explicitly provide a `ranges` field with location information).

klinvill commented:
> 2\. Make sure to await every async function.  (In this case, the `this.handleFullBufferResponse` call in `validateFStarDocument`.)

What does awaiting every async function give us in terms of dealing with unhandled rejections? I think just a broad catch on unhandled rejections should be sufficient to prevent these crashes right?

gebner commented:
> > 3. The full buffer request can also respond with errors apparently:
> 
> Should these error responses result in an error that the IDE highlights? Or just logged to the console? It looks like there is file + line information, but it's buried in a string message that probably won't always occur (as opposed to the `IdeError` messages which explicitly provide a `ranges` field with location information).

The location information in my error message is not from the source file, it's from the location in the F* compiler that raised the error. :smile: 

Ideally, I'd like that error message to appear as a diagnostic on the current location of the full buffer request.  (So that you can see in the editor where F* crashed.)  But for now a console.log is perfectly fine.

gebner commented:
> > 2. Make sure to await every async function.  (In this case, the `this.handleFullBufferResponse` call in `validateFStarDocument`.)
> 
> What does awaiting every async function give us in terms of dealing with unhandled rejections? I think just a broad catch on unhandled rejections should be sufficient to prevent these crashes right?

Our jsonrpc library catches the exception and turns into a jsonrpc error response that is sent back to vscode, as long as we throw it from the request handler (which the await accomplishes).  A catch is totally fine too.  What we should avoid is calling an async function and then doing nothing with the promise.

gebner commented:
@klinvill Are you doing okay over here?  I don't think there's much left to fix the remaining bugs.  If you don't have time, I could finish this as well.

klinvill commented:
Ah sorry about the delay. I got sidetracked with a bunch of other stuff. I can probably have this PR revised and ready for a final review by end-of-day Wednesday. It's also totally fine if you'd prefer to finish it out yourself

gebner commented:
No rush, Kirby!  I have fixed the unhandled rejected promise issue, the code is in my fork: https://github.com/gebner/fstar-vscode-assistant/pull/new/klinvill/async-fstar

@nikswamy If you give me push access to this repo, then I could directly push to Kirby's branch.

klinvill commented:
@gebner I think this PR should be ready now. I merged in the changes in your branch (thanks for making those changes!), added a stronger type for pending_responses, and removed a couple TODOs that had already been addressed. Can you give it one last glance/review?

Also re: permissions, I created this branch in a fork (since I don't have push access) so I think you might have needed push access to my fork to merge in your changes. I merged them in myself, but push access to this repo for you sounds like a good idea to have anyway!

gebner commented:
The last change looks good to me as well.  I'm testing the PR as we speak and I haven't found any issues yet.

> Also re: permissions, I created this branch in a fork (since I don't have push access) so I think you might have needed push access to my fork to merge in your changes. I merged them in myself, but push access to this repo for you sounds like a good idea to have anyway!

There's this nifty "allow edits from maintainers" checkbox that is checked by default when you create a PR.  And that allows anybody with write permission to the target repository to push to the branch in your repository as well.

gebner commented:
We can still improve the request types.  But I think we should merge this PR first to unblock #28 and then tackle the types in another go.

The issue is that the current types imply that all requests are successful.
```typescript
const resp = await conn.lookupQuery(....);
resp.response['defined-at'].fname // type checks, but could fail if resp.status === 'failure'
```
